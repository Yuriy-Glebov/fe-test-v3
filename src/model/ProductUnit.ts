
export const ProductUnit = {
    Kilo: 'Kilo',
    Each: 'Each'
}

export type ProductUnit = typeof ProductUnit[keyof typeof ProductUnit];
