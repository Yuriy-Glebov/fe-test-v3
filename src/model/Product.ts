import {ProductUnit} from "./ProductUnit"

export type Product = {
    name: string;
    unit: ProductUnit;
}

export function createProduct(name: string, unit: ProductUnit): Product {
    return {
      name,
      unit
    }
  }