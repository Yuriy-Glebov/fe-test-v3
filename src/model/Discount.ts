import {Product} from "./Product"

export type Discount = {
  product: Product;
  description: string;
  discountAmount: number;
};

export function createDiscount(product: Product, description: string, discountAmount: number): Discount {
  return {
    product: product,
    description: description,
    discountAmount: discountAmount
  }
}