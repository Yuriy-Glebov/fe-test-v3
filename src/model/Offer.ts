import {Product} from "./Product"
import {SpecialOfferType} from "./SpecialOfferType"

export type Offer = {
    offerType: SpecialOfferType,
    product: Product,
    argument: number,
    getProduct: () => Product,
};

export const createOffer = (offerType: SpecialOfferType, product: Product, argument: number): Offer => {
    return {
        offerType,
        product,
        argument,
        getProduct: () => product,
    };
};
