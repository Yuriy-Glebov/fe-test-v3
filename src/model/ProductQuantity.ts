import {Product} from "./Product"

export type ProductQuantity = {
  product: Product;
  quantity: number;
};

export function createProductQuantity(product: Product, quantity: number): ProductQuantity {
  return {
    product: product,
    quantity: quantity
  }
}