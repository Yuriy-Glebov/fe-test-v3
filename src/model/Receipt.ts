import { Discount } from "./Discount";
import { Product } from "./Product";
import { createReceiptItem, ReceiptItem } from "./ReceiptItem";
import * as _ from "lodash";

export type Receipt = {
  getTotalPrice: () => number;
  addProduct: (
    p: Product,
    quantity: number,
    price: number,
    totalPrice: number
  ) => void;
  getItems: () => ReceiptItem[];
  addDiscount: (discount: Discount) => void;
  getDiscounts: () => Discount[];
};

export function createReceipt() {
  let items: ReceiptItem[] = [];
  let discounts: Discount[] = [];

  return {
    getTotalPrice: function () {
      let total = 0.0;
      for (let item of items) {
        total += item.totalPrice;
      }
      for (let discount of discounts) {
        total -= discount.discountAmount;
      }
      return total;
    },
    addProduct: function (
      p: Product,
      quantity: number,
      price: number,
      totalPrice: number
    ) {
      items.push(createReceiptItem(p, quantity, price, totalPrice));
    },
    getItems: function () {
      return _.clone(items);
    },
    addDiscount: function (discount: Discount) {
      discounts.push(discount);
    },
    getDiscounts: function () {
      return discounts;
    },
  };
}
