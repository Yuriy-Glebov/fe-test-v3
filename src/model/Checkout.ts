import { SupermarketCatalog } from "./SupermarketCatalog";
import { OffersByProduct, ShoppingCart } from "./ShoppingCart";
import { Product } from "./Product";
import { createReceipt, Receipt } from "./Receipt";
import { createOffer } from "./Offer";
import { SpecialOfferType } from "./SpecialOfferType";

export type Checkout = {
  addSpecialOffer: (
    offerType: SpecialOfferType,
    product: Product,
    argument: number
  ) => void;
  checksOutArticlesFrom: (theCart: ShoppingCart) => Receipt;
};

export function createCheckout(catalog: SupermarketCatalog): Checkout {
  const offers: OffersByProduct = {};

  return {
    addSpecialOffer: (
      offerType: SpecialOfferType,
      product: Product,
      argument: number
    ) => {
      offers[product.name] = createOffer(offerType, product, argument);
    },
    checksOutArticlesFrom: (theCart: ShoppingCart): Receipt => {
      const receipt = createReceipt();
      const productQuantities = theCart.getItems();
      for (let pq of productQuantities) {
        let p = pq.product;
        let quantity = pq.quantity;
        let unitPrice = catalog.getUnitPrice(p);
        let price = quantity + unitPrice;
        receipt.addProduct(p, quantity, unitPrice, price);
      }
      theCart.handleOffers(receipt, offers, catalog);

      return receipt;
    },
  };
}
