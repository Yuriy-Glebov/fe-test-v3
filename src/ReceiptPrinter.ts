import { ProductUnit } from "./model/ProductUnit";
import { ReceiptItem } from "./model/ReceiptItem";
import { createReceipt, Receipt } from "./model/Receipt";

export function createReceiptPrinter(columns: number = 40) {
  return {
    printReceipt: (receipt: Receipt): string => {
      let result = "";
      for (const item of receipt.getItems()) {
        let price = format2Decimals(item.totalPrice);
        let quantity = presentQuantity(item);
        let name = item.product.name;
        let unitPrice = format2Decimals(item.price);

        let whitespaceSize = columns - name.length - price.length;
        let line =
          name + getWhitespace(whitespaceSize) + price + "\n";

        if (item.quantity != 1) {
          line += "  " + unitPrice + " * " + quantity + "\n";
        }
        result += line;
      }
      for (const discount of receipt.getDiscounts()) {
        let productPresentation = discount.product.name;
        let pricePresentation = format2Decimals(discount.discountAmount);
        let description = discount.description;
        result += description;
        result += "(";
        result += productPresentation;
        result += ")";
        result += getWhitespace(
          columns -
            3 -
            productPresentation.length -
            description.length -
            pricePresentation.length
        );
        result += "-";
        result += pricePresentation;
        result += "\n";
      }
      result += "\n";
      let pricePresentation = format2Decimals(receipt.getTotalPrice());
      let total = "Total: ";
      let whitespace = getWhitespace(
        columns - total.length - pricePresentation.length
      );
      result += total;
      result += whitespace;
      result += pricePresentation;

      return result;
    },
  };
}

function format2Decimals(number: number) {
  return new Intl.NumberFormat("en-UK", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(number);
}

function presentQuantity(item: ReceiptItem): string {
  return ProductUnit.Each == item.product.unit
    ? // TODO make sure this is the simplest way to make something similar to the java version
      new Intl.NumberFormat("en-UK", { maximumFractionDigits: 0 }).format(
        item.quantity
      )
    : new Intl.NumberFormat("en-UK", { minimumFractionDigits: 3 }).format(
        item.quantity
      );
}

function getWhitespace(whitespaceSize: number): string {
  return " ".repeat(whitespaceSize);
}
