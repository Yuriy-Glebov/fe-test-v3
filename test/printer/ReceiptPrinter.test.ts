import { Product } from "../../src/model/Product";
import { ProductUnit } from "../../src/model/ProductUnit";
import { Receipt } from "../../src/model/Receipt";
import { ReceiptPrinter } from "../../src/ReceiptPrinter";
import { Discount } from "../../src/model/Discount";
import { printStandardReceipt } from "../helper/printStandartReceipt";
const approvals = require("approvals");

type Approvals = { verify: (a: string) => void };

describe("ReceiptPrinter", () => {
  approvals.mocha();

  let toothbrush: Product;
  let apples: Product;
  let receipt: Receipt;

  beforeEach(() => {
    receipt = new Receipt();
    apples = new Product("apples", ProductUnit.Kilo);
    toothbrush = new Product("toothbrush", ProductUnit.Each);
  });

  it("oneLineItem", function (this: any) {
    receipt.addProduct(toothbrush, 1, 0.99, 0.99);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("quantityTwo", function (this: any) {
    receipt.addProduct(toothbrush, 2, 0.99, 0.99 * 2);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("looseWeight", function (this: any) {
    receipt.addProduct(apples, 2.3, 1.99, 1.99 * 2.3);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("total", function (this: any) {
    receipt.addProduct(toothbrush, 1, 0.99, 2 * 0.99);
    receipt.addProduct(apples, 0.75, 1.99, 1.99 * 0.75);
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("discounts", function (this: any) {
    receipt.addDiscount(new Discount(apples, "3 for 2", 0.99));
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });

  it("printWholeReceipt", function (this: any) {
    receipt.addProduct(toothbrush, 1, 0.99, 0.99);
    receipt.addProduct(toothbrush, 2, 0.99, 2 * 0.99);
    receipt.addProduct(apples, 0.75, 1.99, 1.99 * 0.75);
    receipt.addDiscount(new Discount(toothbrush, "3 for 2", 0.99));
    expect(printStandardReceipt(receipt)).toMatchSnapshot();
  });
});
