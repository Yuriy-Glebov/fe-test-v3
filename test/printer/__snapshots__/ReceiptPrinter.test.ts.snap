// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiptPrinter discounts 1`] = `
"3 for 2(apples)                    -0.99

Total:                             -0.99"
`;

exports[`ReceiptPrinter looseWeight 1`] = `
"apples                              4.58
  1.99 * 2.300

Total:                              4.58"
`;

exports[`ReceiptPrinter oneLineItem 1`] = `
"toothbrush                          0.99

Total:                              0.99"
`;

exports[`ReceiptPrinter printWholeReceipt 1`] = `
"toothbrush                          0.99
toothbrush                          1.98
  0.99 * 2
apples                              1.49
  1.99 * 0.750
3 for 2(toothbrush)                -0.99

Total:                              3.47"
`;

exports[`ReceiptPrinter quantityTwo 1`] = `
"toothbrush                          1.98
  0.99 * 2

Total:                              1.98"
`;

exports[`ReceiptPrinter total 1`] = `
"toothbrush                          1.98
apples                              1.49
  1.99 * 0.750

Total:                              3.47"
`;
